### ISSUE TITLE
[BUG] ViewModelViewHost behaves different in 9.4.1 and 9.8.15

### ISSUE THREAD

**Issue by jasonwurzel** [reactiveui/ReactiveUI#1928]
**Describe the bug**

I'm having a simple ViewModelViewHost that gets the ViewModel set and displays the registered View for that ViewModel. The ViewModel in a WhenActivated block in the View is null in version 9.8.15. Previously (in the version I used before updating - 9.4.1) the ViewModel was present. 

**Steps To Reproduce**
1. please see my repo https://github.com/jasonwurzel/rxui-whenactivated-timing
2. start the two projects after one another. 
3. in project rxui_9_4_1 the VM is present by the time the WhenActivated block is executed
5. in project rxui_9_8_15 the VM is null 

**Expected behavior**
Either it was a coincident before and now it is working correctly or the VM should be not null with the latest version as well. My feeling is it should not be null, as the VM should be set before WhenActivated gets called on the fully initialized view.

**Environment(please complete the following information):**
 - OS: Windows 10 1809
 - Version: 9.8.15

**Comment by glennawatson** [comment]
Hi Jason. 

You cannot rely on the WhenActivated to know if the ViewModel exists. 

The WhenActivated was added to handle memory leaks in WPF with WhenAnyValue due to the way Dependency properties work. 

You can use `d(this.WhenAnyValue(x => x.ViewModel).Subscribe(....)`

Also Bind, OneWayBind and those operators also handle changing of view model. 

Relying on the view model already set is dangerous 

Closing the issue due to being as expected but thanks for taking the time to Lodge the issue.

**Comment by jasonwurzel** [comment]
Hi Glenn, 

thank you for the quick response. Ok then I'll correct the places I did it this way.

**Comment by jasonwurzel** [comment]
@glennawatson sorry to bump this issue but I noticed the following behaviour, maybe you can explain it to me: previously (i.e. 9.4.1) I had code in WhenActivated blocks that relied on DependencyProperties being set in xaml, like Color="Black" or alike. When WhenActivated got called, all the Properties were set.
Now though, the properties are NOT yet set. Please see my updated example (I added a "DummyProperty" to rxui_9_8_15.SubView and set it in rxui_9_8_15.MainPage.xaml). 
Again the question is: Have I been doing it wrong all along and should I rely on the View in question to be fully initialized at some other event (Loaded?). Nonetheless this altered behavior seems a pretty big change to me, maybe we should change it back to the way it was before?

**Comment by glennawatson** [comment]
@jasonmurzel I will take a look at it but WhenAnyValue would make sense in this situation too. Reopening for the moment.

**Comment by jasonwurzel** [comment]
heh, it's wurzel, not murzel. As in german for "root" :-D
anyway: I'm having some non-changing properties that get set. I'm shying away from creating a whenany.. for each one of them, as it is absolutely not neccessary. That's why I was wondering if I should resort to using another event.
Thank you for helping out!

**Comment by glennawatson** [comment]
I recently got the uwp unit test project running so this will likely make a good test case. 

The @ completion doesn't work on mobile devices hence the username mistake. I noticed your email name and username were different so didn't want to make the same mistake of doing "Jason" again.

**Comment by jasonwurzel** [comment]
no offense taken! 
by the way, the behavior changed from 9.8.1 to the next version (9.8.9)

**Comment by glennawatson** [comment]
Yeah there are some changes to WhenActivated that fixed bugs with it never being fired in a combobox for example. So it's finding a happy middle ground.

**Comment by jasonwurzel** [comment]
in the ActivationForViewFetcher there's the Loaded Event used, but for UWP the Loading event is used (line 42). Maybe that's the initial cause of the problem? (but it's been that way for 5 years, changed by jlaanstra 
https://github.com/reactiveui/ReactiveUI/commit/688cc8542daab9cc8790624aa2c159462b2a2eab#diff-79f6f1baee10f4eb42599d93680a86daR30 )

**Comment by glennawatson** [comment]
You can register your own ActivationForViewFetcher temporarily if you want to test in splat. Realistically not going to be able to test til this evening.

**Comment by jasonwurzel** [comment]
nice idea, I'll do that

**Comment by jasonwurzel** [comment]
(just writing down the stuff until you have time tonight)
I think I see the difference now: until the latest change, GetActivationForView returned false until IsHitTestVisible returned true:
```csharp 
return viewLoaded
                .Merge(viewUnloaded)
                .Select(b => b ? fe.WhenAnyValue(x => x.IsHitTestVisible).SkipWhile(x => !x) : Observables.False)
                .Switch()
                .DistinctUntilChanged();
```
(in case of UWP, the Loading event immediately returned true (no properties set etc.) but IsHitTestVisible delayed the activation by being false)
now though, in case of UWP, viewLoaded immediately returns true (because it is the FrameworkElement's "Loading" event:
```csharp
            return viewLoaded
                .Merge(viewUnloaded)
                .Merge(hitTestVisible)
                .DistinctUntilChanged();
```
this of course is a big difference between UWP and WPF in ReactiveUI behaviour...one time I get into the WhenActivated block when the View is finished loading, the other time I get into WhenActivated when the view starts loading. It would be interesting why jlaanstra introduced this distinction between WPF and UWP.

**Comment by jasonwurzel** [comment]
so what @moswald said here makes sense: 
`
I just remember that Loading was preferred to Loaded on this platform. IIRC, it's because we can intercept a valid framework element earlier without issue - WhenActivated is only supposed to be used to discover when a control has entered the visual tree, not necessarily when the control is loaded.
`
https://github.com/reactiveui/ReactiveUI/issues/1473#issuecomment-328297155

following this reasoning it works now as expected, as WhenActivated fires much earlier now.
In the end I (seems like maybe others need that too) should implement a `WhenActivationFinished` or alike as @KallynGowdy suggested.

**Comment by glennawatson** [comment]
It's looking likely we should separate off the wpf and uwp view activator.
If you bind to a combolist ItemSource using rxui binding. Does the view
fetching work correctly? That was the issue we changed it for on wpf.

**Comment by jasonwurzel** [comment]
Just did a quick test: The combobox has no issues in either versions ( neither the #1905 bug nor  the #1920 bug)

**Comment by glennawatson** [comment]
Cool. Thanks for confirming that. We will split off uwp/wpf for that class.
Will make the change shortly.

**Comment by jasonwurzel** [comment]
But what should we do with jlaanstra's decision about taking the Loading Event for UWP instead of Loaded, for performance reasons? Basically we could incorporate the #if WINDOWS_UWP part back into the #else part and use "Loaded" for uwp as well. The breaking change @moswald talked about already happened anyway (before, the uwp view was activated when IsHitTestVisible, now when "Loading"). But of course this would result in less than optimal performance for the UWP case...

**Comment by glennawatson** [comment]
It would be the same for uwp as of last week. The difference would be in wpf where comboboxitem don't have IsHitTestVisible so restrict the change to wpf platform.

**Comment by glennawatson** [comment]
The difference is now there will be a wpf one in the reactiveui.wpf project and one in the uap folder of platforms. Take it out of windows-commom

**Comment by glennawatson** [comment]
https://www.nuget.org/packages/reactiveui/9.8.22

**Comment by jasonwurzel** [comment]
thank you!
